os/net/lwip/src/api/sockets.c

```diff
@@ -1835,7 +1861,17 @@ int lwip_poll(int fd, struct pollfd *fds, bool setup)
 
 	sock = tryget_socket(fd);
 	if (!sock) {
-		return -EBADF;
+		/* Need to check select_waiting counter on socket ... */
+		sock = trycheck_selwait_socket(fd);
```

Comment:
Would not it be better to call trycheck_selwait_socket, when the socket deallocation is called (in other thread)? Please let me know your opinion.

Reply:
Thanks for the comment. Actually, de-allocation is in two parts, 1. removing socket netconn structure (which happens in the closing thread), and 2. Decrementing socket select_waiting variable, which happens under poll (poll_teardown specifically). When both are 0, the same socket structure can be re-used. However, at present, part 2 happens after the condition check above. Also, the above condition check is in the right context, i.e. when another thread might not know that the socket fd is closed, and might try to select wait on it. This is where such cases are filtered out. Using the function above in the closing thread will most likely miss this context. Having mentioned so, there might be better alternatives to fixing this, but this is what I gather from the patch.

Reply:
Thanks for the kind reply. Now, I understand the point. Then, what do you think if lwip_poll_teardown is called right after filling zeros in event variables, and then return within this if statement (if (!sock) at line 1863)? It might be more clear to recognize the problem dealt with by this PR happened.